### PR TITLE
Added `max_measurement` field to `Prio3Sum` type

### PR DESCRIPTION
**Pull request by rozbb**
The old `Prio3Sum` let summands anything <2^bits. In newer drafts, the type now has a `max_measurement` parameter that can constrain the size of summands. Closes a subtask of #1122.

Some open questions:
1. Did I get the VDAF metadata correct? I updated `input_len()`, `proof_len()`, and `eval_output_len()` correctly, but what about the other methods? Would I know via unit tests if they were wrong?
2. The new Mastic tests now have to call `input_len()` to find out the correct measurement size to use. Is that an acceptable API for Mastic? I'm not familiar, so this might be fine.

### CODE REVIEW

File: src/vdaf/prio3.rs [hunk 1]
```diff
@@ -1700,11 +1723,14 @@ mod tests {
...
+        let bits = 16;
+        let max_measurement = 35_891;
+
+        let prio3 = Prio3::new_sum(3, bits, max_measurement).unwrap();
 
         assert_eq!(
-            run_vdaf(CTX_STR, &prio3, &(), [0, (1 << 16) - 1, 0, 1, 1]).unwrap(),
-            (1 << 16) + 1
+            run_vdaf(CTX_STR, &prio3, &(), [0, max_measurement - 2, 0, 1, 1]).unwrap(),
+            max_measurement
```
Comment (divergentdave) [inline src/vdaf/prio3.rs:1733]: This would test the edge case of the largest valid measurement.
```suggestion
            run_vdaf(CTX_STR, &prio3, &(), [0, max_measurement, 0, 1, 1]).unwrap(),
            max_measurement + 2
```

File: src/flp/types.rs [hunk 2]
```diff
@@ -154,8 +176,18 @@ impl<F: FftFriendlyFieldElement> Type for Sum<F> {
...
+                "unexpected measurement: got {:?}; want ≤{:?}",
+                summand, self.max_measurement
+            )));
+        }
+
+        let enc_summand: Vec<F> = F::encode_as_bitvector(*summand, self.bits)?.collect();
+        let enc_summand_plus_offset =
+            F::encode_as_bitvector(self.offset + *summand, self.bits)?.collect();
+
+        Ok([enc_summand, enc_summand_plus_offset].concat())
```
Comment (divergentdave) [inline src/flp/types.rs:190]: This will save two allocations and two memcpys:
```suggestion
        let mut encoded = Vec::with_capacity(self.bits * 2);
        let enc_summand = F::encode_as_bitvector(*summand, self.bits)?;
        encoded.extend(enc_summand);
        let enc_summand_plus_offset = F::encode_as_bitvector(self.offset + *summand, self.bits)?;
        encoded.extend(enc_summand_plus_offset);

        Ok(encoded)
```

File: src/flp/types.rs [hunk 3]
```diff
@@ -212,16 +257,16 @@ impl<F: FftFriendlyFieldElement> Type for Sum<F> {
...
     }
 
     fn prove_rand_len(&self) -> usize {
         1
     }
 }
 
-/// The average type. Each measurement is an integer in `[0,2^bits)` for some `0 < bits < 64` and the
-/// aggregate is the arithmetic average.
+/// The average type. Each measurement is a integer in `[0, max_measurement]` and the aggregate is
```
Comment (divergentdave) [inline src/flp/types.rs:268]: nit: copyedit
```suggestion
/// The average type. Each measurement is an integer in `[0, max_measurement]` and the aggregate is
```

**Review by divergentdave — COMMENTED** [review]
FYI, I think this'll need to be rebased to pick up some Clippy lint fixes.

The other methods in the `Type` implementation look good. Most of these get checked by the FLP unit tests or `valid_call_check()`, etc., but maybe not all. I checked them against the definitions in the spec, with formulas using gadget calls, gadget arities, etc.

File: src/flp/types.rs [hunk 4]
```diff
@@ -133,17 +138,34 @@ impl<F: FftFriendlyFieldElement> Debug for Sum<F> {
 
 impl<F: FftFriendlyFieldElement> Sum<F> {
     /// Return a new [`Sum`] type parameter. Each value of this type is an integer in range `[0,
-    /// 2^bits)`.
-    pub fn new(bits: usize) -> Result<Self, FlpError> {
+    /// max_measurement]` where `max_measurement < 2^bits`.
+    ///
+    /// # Panics
+    /// Panics if 2^bits <= max_measurement
+    pub fn new(bits: usize, max_measurement: F::Integer) -> Result<Self, FlpError> {
```
[the remark this answers is not in the document]
Comment (rozbb) [reply src/flp/types.rs:145]: Makes sense, will do

File: src/vdaf/prio3.rs [hunk 5]
```diff
@@ -141,16 +143,29 @@ impl Prio3SumVecMultithreaded {
 pub type Prio3Sum = Prio3<Sum<Field128>, XofTurboShake128, 16>;
 
 impl Prio3Sum {
-    /// Construct an instance of Prio3Sum with the given number of aggregators and required bit
-    /// length. The bit length must not exceed 64.
-    pub fn new_sum(num_aggregators: u8, bits: usize) -> Result<Self, VdafError> {
+    /// Construct an instance of `Prio3Sum` with the given number of aggregators, required bit length,
+    /// and where each summand must be in the range `[0, max_measurement)`. The bit length must not
```
[the remark this answers is not in the document]
Comment (rozbb) [reply src/vdaf/prio3.rs:147]: Oops, ya that's a typo

File: src/flp/types.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (rozbb) [reply src/flp/types.rs]: Done

File: src/vdaf/prio3.rs [hunk 6]
```diff
@@ -141,16 +143,16 @@ impl Prio3SumVecMultithreaded {
...
-                "bit length ({bits}) exceeds limit for aggregate type (64)"
-            )));
-        }
-
-        Prio3::new(num_aggregators, 1, 0x00000001, Sum::new(bits)?)
+    /// Construct an instance of `Prio3Sum` with the given number of aggregators, where each summand
+    /// must be in the range `[0, max_measurement]`.
+    ///
+    /// # Panics
+    /// Panics if `max_measurement == 0`
```
Comment (divergentdave) [inline src/vdaf/prio3.rs:150]: Various panic docs throughout will need to be updated as well

File: src/flp/types.rs [hunk 7]
```diff
@@ -113,37 +113,58 @@ impl<F: FftFriendlyFieldElement> Type for Count<F> {
...
-        }
+    /// max_measurement]` where `max_measurement > 0`.
+    ///
+    /// # Panics
+    /// Panics if `max_measurement == 0`.
+    pub fn new(max_measurement: F::Integer) -> Result<Self, FlpError> {
+        assert!(
+            max_measurement > F::Integer::zero(),
+            "max_measurement must be nonzero"
+        );
```
[the remark this answers is not in the document]
Comment (rozbb) [reply src/flp/types.rs]: Makes sense. Fixed

**Review by divergentdave — APPROVED** [review]
